src/plugins/intel_npu/src/plugin/npuw/just_sync_infer_request.cpp

```diff
@@ -645,19 +725,133 @@ void ov::npuw::JustInferRequest::run_subrequest_for_success(std::size_t idx, boo
+    auto& r = m_subrequests[real_idx];
+    if (!comp_model_desc.spatial) {
+        // Run normally
+        r->infer();
+    } else {
+        // Run over the specified range... Note: the full inputs/outputs
+        // must be prepared in the m_spatial_io at this point
+        const auto& spatial = comp_model_desc.spatial.value();
+        const auto num_outputs = comp_model_desc.compiled_model->outputs().size();
+
+        // Create a sparse vector with full input sizes.
+        // For the access simplicity, its size is aligned with function's
+        // number of input parameters (activations) so some slots may be
+        // not used here.
+        // FIXME: All these preparations could be done statically (just once)
+        std::vector<ov::Shape> full_in_shapes(comp_model_desc.param_base);
+        for (auto&& param : spatial.params) {
+            full_in_shapes[param.idx] = m_spatial_io[real_idx].inputs.at(param.idx)->get_shape();
+        }
+
+        // Now handle the range, even if it is not a multiply of nway (slice):
+        //
+        // |<- - - - full range  - - - ->|
+        // +------+------+------+------+-+
+        // | nway | nway | nway | nway | |
+        // +------+------+------+------+-+
+        //                              ^tail
+        // The block is always compiled to produce nway. If we need a smaller tensor
+        // on the last iteration, the sub-nway will be copied from the input range to
+        // a temporary tensor, and then the sub-nwway range will be copied from the
+        // request's output range.
+
+        std::size_t offset = 0u;
+        for (std::size_t i = 0u; i < spatial.nway_iters; i++, offset += spatial.nway) {
+            // Collect spatial inputs for this offset
+            for (auto&& param : spatial.params) {
+                const auto& iport = comp_model_desc.compiled_model->inputs()[param.idx];
+                r->set_tensor(
+                    iport,
+                    ov::npuw::util::view(m_spatial_io[real_idx].inputs.at(param.idx), param.dim, offset, spatial.nway));
+            }  // for(params)
+
+            // Now set the spatial outputs
+            for (std::size_t out_idx = 0u; out_idx < num_outputs; out_idx++) {
+                const auto& oport = comp_model_desc.compiled_model->outputs()[out_idx];
+                r->set_tensor(oport,
+                              ov::npuw::util::view(m_spatial_io[real_idx].outputs.at(out_idx),
+                                                   spatial.out_dim,
+                                                   offset,
+                                                   spatial.nway));
+            }  // for(outputs)
+
+            // Now run the part
+            r->infer();
+        }  // for(full_nway_times)
+
+        // Now process the tail, if required
+        if (spatial.tail_size) {
+            // Copy the sub-ranges to spatial inputs
+            // NOTE: tails buffers are read from/written to at 0th offset!
```

Comment:
Then why below for params it's `offset` and `0` for outputs? Is it related to the whole sparse representation?

Reply:
tail is a buffer allocated to stencil size. We copy a tail (by definition, less than stencil size) into a dedicated buffer. So we read at offset (from the larger buffer) and write at 0 (in a smaller buffer).